### PR TITLE
arch: riscv: stacktrace: some optimization

### PR DESCRIPTION
**Pull request by ycsin**
1. Stack trace now starts from `mepc`
2. Make sure that the stack unwinds in the right direction.

### CODE REVIEW

**Comment by npitre** [comment]
Strange code flow:

```
        if (esf != NULL) {
                sp = z_riscv_get_sp_before_exc(esf);
                ksp = (uintptr_t *)sp;
                ra = esf->mepc;
        } else {
                return;
        }
```

Not wrong but strange.

Is there a reason why you didn't do this instead?

```
        if (esf == NULL) {
                return;
        }

        sp = z_riscv_get_sp_before_exc(esf);
        ksp = (uintptr_t *)sp;
        ra = esf->mepc;
```

**Comment by ycsin** [comment]
> Is there a reason why you didn't do this instead?

The implementation here is kinda extracted from #73587, so it retained the pattern over there.

File: arch/riscv/core/stacktrace.c [hunk 1]
```diff
@@ -88,15 +88,20 @@ static inline bool in_text_region(uintptr_t addr)
 #ifdef CONFIG_FRAME_POINTER
 void z_riscv_unwind_stack(const struct arch_esf *esf)
 {
-	uintptr_t fp = esf->s0;
+	uintptr_t fp, last_fp = 0;
 	uintptr_t ra;
 	struct stackframe *frame;
 
+	if (esf != NULL) {
```
Comment (kgugala) [inline arch/riscv/core/stacktrace.c:95]: why did you move this check to the `unwind` function? Before, this was checked before calling the function, now we have to have this check in two places - as there are two implementations of `z_riscv_unwind_stack`

**Review by kgugala — CHANGES_REQUESTED** [review]
